### PR TITLE
ARROW-8601: [Go][Flight] Implementations Flight RPC server and client

### PR DESCRIPTION
**Pull request by zeroshade**
Built out from https://github.com/apache/arrow/pull/6731 with some inspiration from the existing Reader/Writer and the C++ Flight implementation. Still need to build out the tests some more, but would like to get opinions and thoughts on what I've got so far as I continue to build it out.

### CODE REVIEW

**Comment by zeroshade** [comment]
Tagging the people that were on the previous PR for visibility:

@sbinet @mindhash @stuartcarnie could any / all of you have a look / review? Thanks much!

**Comment by zeroshade** [comment]
Bump. Can someone please give this a review so we can get this merged? Thanks.

**Comment by wesm** [comment]
Agree it would be good for this to be reviewed / merged. Can you rebase and get a more complete CI build?

**Comment by zeroshade** [comment]
@wesm rebased

**Comment by zeroshade** [comment]
@wesm as far as i can tell the two checks that are failing are unrelated to this PR. 

I have ideas for further exploring / adding more functionality for the FlightRPC stuff, but I don't want to put too much into this PR and would rather further development be a separate PR to reach more feature parity with the C++ / Python versions of the flight server/client. I'm actually currently building out a project at work that will be utilizing the Go flight rpc server and client which was the impetus for me building this out. So anything you can do to help get this reviewed and pushed along would be greatly appreciated. 

Thanks much!

**Review by sbinet — COMMENTED** [review]
just a quick, drive-by, review.

I apologize, but I've switched teams at work and my new position doesn't involve much using Arrow, so that's only free cycles.
sorry.

File: go/arrow/flight/client.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flight
+
+import (
+	context "context"
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply go/arrow/flight/client.go:20]: i don't know why my IDE did that -_- i'll fix that.

**Comment by zeroshade** [comment]
@wesm @sbinet @stuartcarnie @mindhash bump

**Review by wesm — APPROVED** [review]
+1. I will go ahead and merge since CI is passing. Thanks @zeroshade!

**Comment by stuartcarnie** [comment]
Yes, thanks!

**Comment by kszucs** [comment]
```
apache-rat license violation: go/arrow/flight/Flight_grpc.pb.go
apache-rat license violation: go/arrow/flight/example_flight_server_test.go
```

These files should have been added to the `rat_exclude_files.txt`.

**Pull request closed without merging**